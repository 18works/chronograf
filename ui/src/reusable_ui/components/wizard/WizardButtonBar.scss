/*
  Wizard Button Bar Styles
  ------------------------------------------------------------------------------
*/

@import 'src/style/modules/modules';

.wizard-button-bar {
  display: flex;
  margin: $ix-marg-b auto $ix-marg-c auto;
  position: relative;
  min-width: 100%;
  justify-content: center;
  align-items: center;
  height: 50px;
  .btn {
    position: relative;
    white-space: nowrap;
    margin: 0 $ix-marg-a;
  }
}