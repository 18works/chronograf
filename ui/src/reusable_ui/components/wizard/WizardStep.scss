@import 'src/style/modules/variables';

.wizard-step-content {
  margin: $ix-marg-b 0;
}

.wizard-step-child {
  padding: 0 $ix-marg-c;
}