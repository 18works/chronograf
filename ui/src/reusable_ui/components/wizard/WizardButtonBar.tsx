import React, {PureComponent} from 'react'

import 'src/reusable_ui/components/wizard/WizardButtonBar.scss'

interface Props {
  decrement?: () => void
  nextLabel?: string
  previousLabel?: string
  lastStep?: boolean
  onClickPrevious: () => void
  onClickNext: () => void
}

class WizardButtonBar extends PureComponent<Props> {
  public static defaultProps: Partial<Props> = {
    nextLabel: 'next',
    previousLabel: 'previous',
  }

  public render() {
    const {
      decrement,
      previousLabel,
      nextLabel,
      onClickPrevious,
      onClickNext,
    } = this.props
    return (
      <div className="button-bar">
        {decrement && (
          <button className="btn btn-md btn-default" onClick={onClickPrevious}>
            {previousLabel}
          </button>
        )}
        <button
          className={`btn btn-md ${this.buttonColor}`}
          onClick={onClickNext}
        >
          {nextLabel || 'next'}
        </button>
      </div>
    )
  }

  private get buttonColor() {
    const {lastStep} = this.props

    if (lastStep) {
      return 'btn-success'
    }

    return 'btn-primary'
  }
}

export default WizardButtonBar
