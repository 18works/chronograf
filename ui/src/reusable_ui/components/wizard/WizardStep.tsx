import React, {PureComponent, ReactNode} from 'react'

import 'src/reusable_ui/components/wizard/WizardStep.scss'

interface Props {
  children: ReactNode
  title: string
  isComplete: () => boolean
  onPrevious: () => void
  onNext: () => void
  increment?: () => void
  decrement?: () => void
  tipText?: string
  nextLabel?: string
  previousLabel?: string
  lastStep?: boolean
}

class WizardStep extends PureComponent<Props> {
  public render() {
    const {children, decrement, tipText, nextLabel, previousLabel} = this.props

    return (
      <div className="progress-step">
        <div className="tip-text">
          <p>{tipText}</p>
        </div>
        <div className="step-content">{children}</div>
        <div className="button-bar">
          {decrement && (
            <button
              className="btn btn-md btn-default"
              onClick={this.handleClickPrevious}
            >
              {previousLabel || 'back'}
            </button>
          )}
          <button
            className={`btn btn-md ${this.buttonColor}`}
            onClick={this.handleClickNext}
          >
            {nextLabel || 'next'}
          </button>
        </div>
      </div>
    )
  }

  private handleClickPrevious = () => {
    const {onPrevious, decrement} = this.props
    onPrevious() // TODO wait if async function
    decrement()
  }

  private handleClickNext = () => {
    const {onNext, increment} = this.props
    onNext()
    increment()
  }

  private get buttonColor() {
    const {lastStep} = this.props

    if (lastStep) {
      return 'btn-success'
    }

    return 'btn-primary'
  }
}

export default WizardStep
