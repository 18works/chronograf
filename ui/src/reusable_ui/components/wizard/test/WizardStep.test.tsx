import React from 'react'
import {shallow} from 'enzyme'

import WizardStep from 'src/reusable_ui/components/wizard/WizardStep'
import FancyScrollbar from 'src/shared/components/FancyScrollbar'
import WizardButtonBar from 'src/reusable_ui/components/wizard/WizardButtonBar'

describe('WizardStep', () => {
  let wrapper

  const mockChild = 'this is a text'

  const decrement = jest.fn()

  const increment = jest.fn()

  const props = {
    title: 'my wizard step',
    isComplete: () => true,
    onPrevious: jest.fn(),
    onNext: jest.fn(),
    increment: undefined,
    decrement: undefined,
    tipText: undefined,
    nextLabel: undefined,
    previousLabel: undefined,
    lastStep: undefined,
  }

  beforeEach(() =>
    (wrapper = shallow(<WizardStep {...props}>{mockChild}</WizardStep>)))

  it('mounts without exploding', () => {
    expect(wrapper).toHaveLength(1)
  })

  it('mounts a fancy scrollbar', () => {
    expect(wrapper.find(FancyScrollbar)).toHaveLength(1)
  })

  it('mounts a button bar', () => {
    expect(wrapper.find(WizardButtonBar)).toHaveLength(1)
  })

  it('mounts children', () => {
    expect(wrapper.find(FancyScrollbar).props().children).toBe(mockChild)
  })

  describe('WizardStep handleClickPrevious', () => {
    const newProps = {...props, decrement}

    beforeEach(() => {
      jest.resetAllMocks()
      wrapper = shallow(<WizardStep {...newProps}>{mockChild}</WizardStep>)
    })

    it('calls onPrevious on handleClickPrevious', () => {
      const spy = jest.spyOn(newProps, 'onPrevious')

      expect(spy).not.toBeCalled()
      wrapper.instance().handleClickPrevious()
      expect(spy).toBeCalled()
    })

    it('calls decrement on handleClickPrevious', async () => {
      const spy = jest.spyOn(newProps, 'decrement')

      expect(spy).not.toBeCalled()
      await wrapper.instance().handleClickPrevious()
      expect(spy).toBeCalled()
    })
  })

  describe('WizardStep handleClickNext', () => {
    const newProps = {...props, increment}

    beforeEach(() => {
      jest.resetAllMocks()
      wrapper = shallow(<WizardStep {...newProps}>{mockChild}</WizardStep>)
    })

    it('calls onNext on handleClickNext', () => {
      const spy = jest.spyOn(newProps, 'onNext')

      expect(spy).not.toBeCalled()
      wrapper.instance().handleClickNext()
      expect(spy).toBeCalled()
    })

    it('calls increment on handleClickNext', async () => {
      const spy = jest.spyOn(newProps, 'increment')

      expect(spy).not.toBeCalled()
      await wrapper.instance().handleClickNext()
      expect(spy).toBeCalled()
    })
  })
})
