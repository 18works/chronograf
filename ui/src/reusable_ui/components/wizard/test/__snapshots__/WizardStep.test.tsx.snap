// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WizardStep matches snapshot 1`] = `
<React.Fragment>
  <FancyScrollbar
    autoHeight={true}
    autoHide={false}
    className="wizard-step-content"
    maxHeight={400}
    setScrollTop={[Function]}
    style={Object {}}
  >
    <div
      className="wizard-step-child"
    >
      this is a text
    </div>
  </FancyScrollbar>
  <WizardButtonBar
    nextLabel="next"
    onClickNext={[Function]}
    onClickPrevious={[Function]}
    previousLabel="previous"
  />
</React.Fragment>
`;
